File: C++/21. Merge Two Sorted Lists.cpp
// Iterative
ListNode* mergeTwoLists(ListNode* l1, ListNode* l2) 
{
    if(l1 == NULL)
        return l2 ; 

    if(l2 == NULL)
        return l1 ; 

    ListNode *head, *tail ; 

    if(l1->val < l2->val)
    {
        head = l1 ; 
        tail = l1 ; 
        l1 = l1->next ; 
    }
    else
    {
        head = l2 ; 
        tail = l2 ;
        l2 = l2->next ; 
    }

    while(l1 and l2)
    {
        if(l1->val < l2->val)
        {
            tail->next = l1 ; 
            l1 = l1->next ; 
        }
        else
        {
            tail->next = l2 ; 
            l2 = l2->next ; 
        }
        tail = tail->next ;
    }

    if(l1)
        tail->next = l1 ; 
    
    if(l2)
        tail->next = l2 ; 

    return head ; 
}

// Recursive Approach
ListNode* mergeTwoLists(ListNode* l1, ListNode* l2) 
{
    if(l1 == NULL)
        return l2 ; 

    if(l2 == NULL)
        return l1 ; 

    if(l1->val < l2->val)
    { 
        l1->next = mergeTwoLists(l1->next, l2);
        return l1 ;   
    }
    else
    { 
        l2->next = mergeTwoLists(l1, l2->next);
        return l2 ; 
    }
}

File: Java/21. Merge Two Sorted Lists.java
class Solution {

    public ListNode mergeTwoLists(ListNode list1, ListNode list2) {
        ListNode head = null, cur = null;

        while (list1 != null && list2 != null) {
            if (list1.val < list2.val) {
                if (cur == null) {
                    head = list1;
                    cur = head;
                } else {
                    cur.next = list1;
                    cur = cur.next;
                }
                list1 = list1.next;
            } else {
                if (cur == null) {
                    head = list2;
                    cur = head;
                } else {
                    cur.next = list2;
                    cur = cur.next;
                }
                list2 = list2.next;
            }
        }

        if (list1 != null) {
            if (cur == null)
                head = list1;
            else
                cur.next = list1;
        }

        if (list2 != null) {
            if (cur == null)
                head = list2;
            else
                cur.next = list2;
        }

        return head;
    }

    public class ListNode {
        public int val;
        public ListNode next;

        public ListNode() {
        }

        public ListNode(int val) {
            this.val = val;
        }

        public ListNode(int val, ListNode next) {
            this.val = val;
            this.next = next;
        }
    }
}